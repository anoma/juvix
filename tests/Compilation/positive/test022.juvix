-- lists
module test022;

open import Stdlib.Prelude;
open import Stdlib.Data.Nat.Ord;

gen : Nat → List Nat;
gen zero := nil;
gen n@(suc m) := n :: gen m;

sum : Nat → Nat;
sum n := foldl (+) 0 (gen n);

sum' : Nat → Nat;
sum' n := foldr (+) 0 (gen n);

printListNatLn : List Nat → IO;
printListNatLn nil := printStringLn "nil";
printListNatLn (h :: t) := printNat h >> printString " :: " >> printListNatLn t;

main : IO;
main :=
  printListNatLn (gen 10) >>
  printListNatLn (reverse (gen 10)) >>
  printListNatLn (filter ((<) 5) (gen 10)) >>
  printListNatLn (reverse (map (flip sub 1) (gen 10))) >>
  printNatLn (sum 10000) >>
  printNatLn (sum' 10000);

end;
