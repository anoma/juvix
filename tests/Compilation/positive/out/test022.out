10 :: 9 :: 8 :: 7 :: 6 :: 5 :: 4 :: 3 :: 2 :: 1 :: nil
1 :: 2 :: 3 :: 4 :: 5 :: 6 :: 7 :: 8 :: 9 :: 10 :: nil
10 :: 9 :: 8 :: 7 :: 6 :: nil
0 :: 1 :: 2 :: 3 :: 4 :: 5 :: 6 :: 7 :: 8 :: 9 :: nil
50005000
50005000
